commons-mfa/src/main/java/org/exoplatform/mfa/filter/MfaFilter.java

```diff
@@ -28,18 +30,18 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
     HttpSession session = httpServletRequest.getSession();
     ExoContainer container = PortalContainer.getInstance();
     MfaService mfaService = container.getComponentInstanceOfType(MfaService.class);
-  
-    String requestUri = httpServletRequest.getRequestURI();
-    if (mfaService.isProtectedUri(requestUri)) {
-      if (session.getAttribute("mfaValidated")==null || !(boolean)session.getAttribute("mfaValidated")) {
-        LOG.info("Mfa Filter must redirect on page to fill token");
-        httpServletResponse.sendRedirect("/portal/dw/mfa-access?initialUri="+requestUri);
-        return;
+    UserACL userACL = CommonsUtils.getService(UserACL.class);
+    String[] protectedGroup = System.getProperty("exo.mfa.protectedGroupNavigations") != null ? System.getProperty("exo.mfa.protectedGroupNavigations").split(",") : new String[0];
+    for (String group : protectedGroup) {
+      String requestUri = httpServletRequest.getRequestURI();
+      if (mfaService.isProtectedUri(requestUri) && userACL.isUserInGroup(group)) {
+        if (session.getAttribute("mfaValidated") == null || !(boolean) session.getAttribute("mfaValidated")) {
```

Comment:
The check of mfaValidated should be before the check of groups (it will have better performances : no need to compute if user can access is he already have mfaValidated)